kaldb/src/main/java/com/slack/kaldb/logstore/search/KaldbDistributedQueryService.java

```diff
@@ -126,6 +131,67 @@ private KaldbServiceGrpc.KaldbServiceFutureStub getKaldbServiceGrpcClient(String
         .withCompression("gzip");
   }
 
+  @VisibleForTesting
+  public static Collection<String> getSnapshotsToSearch(
+      SnapshotMetadataStore snapshotMetadataStore,
+      SearchMetadataStore searchMetadataStore,
+      long startTimeMs,
+      long endTimeMs) {
+    // step 1 - find all snapshots that match time window
+    Set<String> snapshotsToSearch =
+        snapshotMetadataStore
+            .getCached()
+            .stream()
+            .filter(
+                snapshotMetadata ->
+                    containsDataInTimeRange(
+                        snapshotMetadata.startTimeEpochMs,
+                        snapshotMetadata.endTimeEpochMs,
+                        startTimeMs,
+                        endTimeMs))
+            .map(snapshotMetadata -> snapshotMetadata.name)
+            .collect(Collectors.toSet());
+
+    // TODO: in the future, we'd have another field in search metadata to indicate if it's hosted by
+    // an indexer or cache node and
+    // it won't be a shuffle, but can be something more nuanced so that we prefer cache over indexer
+
+    // step 2 - iterate every snapshot and map it to the search metadata while de-duplicating search
+    // metadata with the same snapshot name
+    return searchMetadataStore
+        .getCached()
+        .stream()
+        .filter(searchMetadata -> snapshotsToSearch.contains(searchMetadata.snapshotName))
+        // ensure we randomize the list so that when there are multiple
+        // search metadata nodes with the same snapshot we don't always pick the same one
+        .collect(
+            Collectors.toMap(
+                SearchMetadata::getSnapshotName,
+                SearchMetadata::getUrl,
+                (firstUrl, secondUrl) -> random.nextBoolean() ? firstUrl : secondUrl))
```

Comment:
I had done that as the first approach. But that required a deep copy of the list so I thought this might be better off.

I can revert to shuffling like I had in this commit - https://github.com/slackhq/kaldb/pull/235/commits/eebbc330388b55820100afa34061831f419a176b

Reply:
> .filter(...)
> .groupBy(snapshotName)
> .map(list -> list.size() > 1 ? list.get(0) : list.get(random(id)))
> .collect()